src/garage/sampler/ray_sampler.py

```diff
@@ -9,95 +9,125 @@
 from collections import defaultdict
 import pickle
 
-import numpy as np
 import psutil
 import ray
 
-from garage.experiment import deterministic
+from garage import TrajectoryBatch
 from garage.misc.prog_bar_counter import ProgBarCounter
-from garage.sampler.base import BaseSampler
+from garage.sampler.sampler import Sampler
 
 
-class RaySampler(BaseSampler):
+class RaySampler(Sampler):
     """Collects Policy Rollouts in a data parallel fashion.
 
     Args:
-        algo (garage.np.algo.RLAlgorithm): A garage algo object
-        env (gym.Env): A gym/akro env object
-        seed (int): Random seed.
-        should_render (bool): If the sampler render the trajectories.
-        num_processors (int): Number of processors to be used.
-        sampler_worker_cls (garage.sampler.ray_sampler.SamplerWorker):
-            If none, uses the default SamplerWorker class
+        worker_factory(garage.sampler.WorkerFactory): Used for worker behavior.
+        agents(list[garage.Policy]): Agents to distribute across workers.
+        envs(list[gym.Env]): Environments to distribute across workers.
+        num_processors(int): Number of workers processes to spawn. If none,
+            defaults to number of physical CPU cores.
+        sampler_worker_cls(type): If None, uses the default SamplerWorker
+            class.
 
     """
 
     def __init__(self,
-                 algo,
-                 env,
-                 seed,
-                 should_render=False,
+                 worker_factory,
+                 agents,
+                 envs,
                  num_processors=None,
                  sampler_worker_cls=None):
-        super().__init__(algo, env)
+        # pylint: disable=super-init-not-called
         self._sampler_worker = ray.remote(SamplerWorker if sampler_worker_cls
                                           is None else sampler_worker_cls)
-        self._seed = seed
-        deterministic.set_seed(self._seed)
-        self._max_path_length = self.algo.max_path_length
-        self._should_render = should_render
+        self._worker_factory = worker_factory
+        self._agents = agents
+        self._envs = self._worker_factory.prepare_worker_messages(envs)
         if not ray.is_initialized():
             ray.init(log_to_driver=False)
         self._num_workers = (num_processors if num_processors else
                              psutil.cpu_count(logical=False))
         self._all_workers = defaultdict(None)
-        self._idle_worker_ids = list(range(self._num_workers))
-        self._active_worker_ids = []
+        self._workers_started = False
+        self.start_worker()
+
+    @classmethod
+    def from_worker_factory(cls, worker_factory, agents, envs):
```

Comment:
I don't quite understand why we need an additional function for constructing the sampler other than the constructor of the sampler itself. @krzentner can you please share your thoughts on this design choice

Reply:
if you have this feedback, please put it in the PR which introduces this API, not this PR (which merely refactors to use TrajectoryBatch).

Reply:
Basically the rationale is: `__init__` in interfaces is unusual and current implementations of `__init__` are all different.